### PR TITLE
rustc: Enable embedding LLVM bitcode for iOS

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit updates rustc to embed bitcode in each object file generated by
default when compiling for iOS. This was determined in #35968 as a step
towards better compatibility with the iOS toolchain, so let's give it a spin and
see how it turns out!

Note that this also updates the `cc` dependency which should propagate this
change of embedding bitcode for C dependencies as well.

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @comex, @michaeleiselsc, y'all may want to double check me on this!

**Comment by alexcrichton** [comment]
Some helpful links thanks to @comex 

* https://github.com/llvm-mirror/clang/blob/b641d31365414ba3ea0305fdaa80369a9efb6bd9/lib/CodeGen/BackendUtil.cpp#L1242 - LLVM's support for this
* https://opensource.apple.com/source/ld64/ld64-274.2/src/ld/parsers/macho_relocatable_file.cpp.auto.html - the linker source where these sections are loaded

**Comment by michaeleiselsc** [comment]
i'm still not sure if we want to always have bitcode on, i think it may cause binary bloat for internal distributions, @comex do you know?

**Comment by comex** [comment]
Well, I don't know of anything that would normally strip it out.  (In particular, `strip` doesn't remove bitcode, which makes sense because Xcode runs `strip` by default when building the archive you'd send to Apple.)  I don't know whether any third-party app distribution platforms (for betas or internal corporate apps) have custom tooling to strip it out.  I also don't know how much bitcode typically adds to binary sizes.  So… no idea :)

I'm building this PR locally and will at least do a basic test of whether I can build a Rust static library and link it into an iOS app with bitcode enabled, without the linker complaining about missing bitcode objects.  That wouldn't guarantee that the bitcode is actually reasonable: in particular, rustc currently defaults to building against LLVM 6, while the latest Xcode is based on LLVM 5, and I have no idea what Apple's servers are running.  But it's a start.

**Comment by michaeleiselsc** [comment]
a few more questions:
- what happens when the linker receives a mix of bitcode files and non-bitcode files, e.g. because a project doesn't want to use bitcode but is built with a mix of rust and other languages where bitcode is turned off?
- does the app store assume that bitcode in the executable means that it should use the bitcode for its distribution process, or is there another toggle the user has to use? if it's the former, then we would be forcing people to use the app store bitcode process

**Comment by alexcrichton** [comment]
@comex thanks for testing!

@michaeleiselsc I'm not so sure myself, I'm relying on y'all for guidance on what you need from rustc :(

**Comment by comex** [comment]
I tried to test it but ran into some issues with the standard library build - in particular, #48906, which I just filed, combined with an odd bug where `nm` (and maybe other things?) rejects object files that were compiled with `-fembed-bitcode-marker`.  I'll look further tomorrow.

**Comment by michaeleiselsc** [comment]
It feels like there are a lot of potential issues with always generating bitcode, so although I can take a look at them, I feel like we should provide an option not to emit bitcode. In fact, because bitcode in Xcode is off by default IIRC, perhaps we should make the default be not to generate bitcode and just provide this as an option

**Comment by comex** [comment]
Okay, I figured out what's going on.

- LLVM's code for reading object files has a bug that causes it to mishandle `-fembed-bitcode-marker` objects.  It's a pretty obvious bug - e.g. you get an error if you run `nm` on such objects - but I guess nobody noticed because `-fembed-bitcode-marker` has only been used by Xcode in limited circumstances.  [I submitted a patch to LLVM upstream.](https://reviews.llvm.org/D44373)  (This does not affect objects with actual embedded bitcode.)
- rustc's bootstrap has a bug (filed as #48906) that causes it to build libbacktrace (among other things) with `-fembed-bitcode-marker` even when optimization is on, which I filed as #48906.  Not to be confused with an entirely separate bug I discovered during the process, #48903, with a different cause, that causes libbacktrace to actually be built without optimization (on all platforms!) :)
- The aforementioned LLVM bug causes `rustc` to produce broken static libraries: specifically, the index (as shown by `nm --print-armap`) is missing all symbols from objects compiled with `-fembed-bitcode-marker`.  If you extract the object files from the archive, the symbols are there, but they're not indexed.  (As far as I can tell, this issue doesn't affect libraries created with `ar`, which makes sense since Apple still ships `ar` from `cctools` rather than symlinking it to `llvm-ar`.  But `rustc` produces static libraries directly using LLVM, rather than using `ar`.)
- So on my test setup, which has rustc make a static library for Xcode to link to, `ld` complains about undefined symbols for `_backtrace_create_state`, `_backtrace_syminfo`, `_backtrace_pcinfo`.  (This is the exact same error @michaeleiselsc encountered [in this post](https://github.com/rust-lang/rust/issues/35968#issuecomment-369105046), but in his case there was an entirely different cause: he was using `--emit=llvm-bc`, so libbacktrace wasn't being included at all.)
- As a hack, if I run `ranlib` on the library before linking, to fix the index, my test setup works!  But of course that should not be necessary.

For now, as a workaround for the LLVM bug, I think the best approach is to avoid `-fembed-bitcode-marker` or equivalent, in both this PR and `cc-rs`.  Either embed bitcode by default even for debug builds, or just disable entirely by default for now.  Once the LLVM bug has been fixed and the fix is shipping in Xcode, the default can be changed back.

Also, I agree with @michaeleiselsc that there should be an option to control the behavior.

**Comment by michaeleiselsc** [comment]
@comex does this relate to the dsymutil segfault I was getting? It seems that rustc does not always generate good symbols (#46447), so perhaps this is triggering that

**Comment by alexcrichton** [comment]
Thanks for the investigation @comex! I've updated to fix those various rustbuild issues you pointed out and went ahead and disabled this by default for iOS. It can be enabled with `-Z embed-bitcode`

**Comment by michaelwoerister** [comment]
Ping me when this is ready for review.

**Comment by alexcrichton** [comment]
Ok @michaelwoerister I think this should be ready!

@michaeleiselsc errors like https://github.com/rust-lang/rust/issues/46447 *may* have been fixed by https://github.com/rust-lang/rust/pull/46772, although maybe not as well

File: src/librustc_trans/back/write.rs [hunk 1]
```diff
@@ -796,6 +819,59 @@ unsafe fn codegen(cgcx: &CodegenContext,
...
+        "rustc.embedded.module\0".as_ptr() as *const _,
+    );
+    llvm::LLVMSetInitializer(llglobal, llconst);
+    let section = if cgcx.opts.target_triple.contains("-ios") {
+        "__LLVM,__bitcode\0"
+    } else {
+        ".llvmbc\0"
+    };
+    llvm::LLVMSetSection(llglobal, section.as_ptr() as *const _);
+    llvm::LLVMRustSetLinkage(llglobal, llvm::Linkage::PrivateLinkage);
```
Comment (michaelwoerister) [inline src/librustc_trans/back/write.rs]: I guess this is what clang is doing?

File: src/libstd/build.rs [hunk 2]
```diff
@@ -86,6 +86,7 @@ fn main() {
 
 fn build_libbacktrace(host: &str, target: &str) -> Result<(), ()> {
     let native = native_lib_boilerplate("libbacktrace", "libbacktrace", "backtrace", ".libs")?;
+    let cflags = env::var("CFLAGS").unwrap_or_default() + " -fvisibility=hidden -O2";
```
Comment (michaelwoerister) [inline src/libstd/build.rs]: Is this related to bitcode embedding?

**Comment by michaelwoerister** [comment]
Thanks, @alexcrichton! Looks good to me. 

@bors r+

File: src/libstd/build.rs (hunk 2 above)
Comment (alexcrichton) [reply src/libstd/build.rs]: Sort of tangentially yeah, but it's otherwise fixing issues that @comex brought up when testing this

Comment (michaelwoerister) [reply src/libstd/build.rs]: I later saw that this is in a separate commit with proper commit message anyway. Thanks for the info!
